nni/algorithms/compression/v2/pytorch/utils/evaluator.py

```diff
@@ -0,0 +1,581 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from __future__ import annotations
+
+from copy import deepcopy
+import logging
+import types
+from typing import Dict, List, Tuple, Union, Any, Callable, Optional
+
+from torch import Tensor
+from torch.nn import Module
+from torch.optim import Optimizer
+from torch.optim.lr_scheduler import _LRScheduler
+from torch.utils.hooks import RemovableHandle
+
+import pytorch_lightning as pl
+from pytorch_lightning.callbacks import Callback
+
+from nni.common import is_traceable
+from .constructor_helper import OptimizerConstructHelper, LRSchedulerConstructHelper
+
+_logger = logging.getLogger(__name__)
+
+
+class Hook:
+    """
+    The base class used to generate, register and remove torch hook.
+
+    Parameters
+    ----------
+    target
+        The hook target, a torch.Tensor or a torch.nn.Module.
+    target_name
+        The name of the target, use periods to separate, e.g., 'model.layers.0.conv1.weight'.
+    hook_factory
+        A factory fucntion, input is an empty list, output is a hook function.
+        The empty list is used to store some useful information in hook.
+    """
+
+    def __init__(self, target: Module | Tensor, target_name: str, hook_factory: Callable[[List], Callable]):
+        self.target = target
+        self.target_name = target_name
+        self.hook_factory = hook_factory
+        self.buffer: List = []
+        self.handle: RemovableHandle | None = None
+
+    def _register(self, hook_func: Callable) -> RemovableHandle:
+        raise NotImplementedError
+
+    def register(self):
+        if self.handle is not None:
+            _logger.warning('%s for %s already has been registered.', self.__class__.__name__, self.target_name)
+            return
+        self.handle = self._register(self.hook_factory(self.buffer))
+
+    def remove(self):
+        if self.handle is None:
+            _logger.warning('%s for %s has not been registered yet.', self.__class__.__name__, self.target_name)
+            return
+        self.handle.remove()
+        self.handle = None
+        self.buffer = []
+
+
+class TensorHook(Hook):
```

Comment:
What is a tensor hook?

Reply:
use to hook the tensor grad https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html

Reply:
Why do we have a base class called Hook? What will happen if users write a customized hook following the tutorial you provided above?

Reply:
This is an internal util class, use to save some hook information (name, buffer, handle...), I think users will never this

Reply:
will add some examples in docstrings for these `XXHook`